### PR TITLE
datasource: add vmware backdoor

### PR DESCRIPTION
**Pull request by crawford**

### CODE REVIEW

**Comment by crawford** [comment]
Updated to include DNS servers and use sigma/vmx-guestinfo.

File: Documentation/vmware-backdoor.md [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+| `hostname`                        | `hostname`                      |
+| `interfaces.<n>.ipv4.<m>.proto`   | `{"dhcp", "static"}`            |
+| `interfaces.<n>.ipv4.<m>.address` | `IPv4 address`                  |
+| `interfaces.<n>.ipv4.<m>.netmask` | `IPv4 netmask`                  |
+| `interfaces.<n>.ipv6.<m>.proto`   | `{"dhcp", "static"}`            |
+| `interfaces.<n>.ipv6.<m>.address` | `IPv6 address`                  |
+| `interfaces.<n>.ipv6.<m>.cidr`    | `int`                           |
+| `interfaces.<n>.role`             | `{"public", "private"}`         |
+| `interfaces.<n>.mac`              | `MAC address`                   |
+| `interfaces.<n>.gateway`          | `IP address`                    |
```
Comment (marineam) [inline Documentation/vmware-backdoor.md]: Well that's just depressing that the whoever came up with this list didn't consider both v4 and v6 or other routes but ok. I suppose it is assumed that v6 always uses router discovery even if the address itself is static.

**Comment by marineam** [comment]
LGTM

File: Documentation/vmware-backdoor.md (hunk 1 above)
Comment (crawford) [inline Documentation/vmware-backdoor.md]: It's not too late to change this! In fact, we should do it now if we are going to make the change. Are you suggesting that it read `interfaces.<n>.gateway.<m>`, or is that not expressive enough (e.g. declaring multiple routes)?

**Comment by sigma** [comment]
is there anything I can do to make this move forward? Up until now, I'm still maintaining my fork of CoreOS just for the ability to get the cloud-config file through the backdoor :)
https://github.com/sigma/coreos-cloudinit/tree/t/feature/vmw-init

Note that I updated sigma/vmw-guestinfo to realign it with the code base of open-vm-tools 9.10.2

**Comment by crawford** [comment]
@sigma We were waiting a vendor to finish (begin) testing but I've about had it. This will be in next week's Alpha. Sorry for the delay.

**Comment by sigma** [comment]
@crawford awesome, thanks. I'm gonna ditch my branch and build on top of yours then.
From a quick look, it doesn't look like guestinfo.coreos.config can be a URL for example (which I find convenient in my branch :)). Any particular reason?

**Comment by sigma** [comment]
@crawford btw, if said vendor is VMware, I can volunteer some of my/their time to do the testing ;)

**Comment by crawford** [comment]
@sigma that's a decent idea. What if we add `guestinfo.coreos.config.url`?

And it's not VMware :)

**Comment by crawford** [comment]
@sigma also, FYI, I'm going to rework the options slightly so they aren't so archaic

**Comment by sigma** [comment]
@crawford sure, guestinfo.coreos.config.url sounds just fine.

**Comment by crawford** [comment]
Uhhhh...

```
vmware_test.go:293: bad error (#5): want invalid CIDR address: test destination, got invalid CIDR address: test destination
```

**Comment by crawford** [comment]
@sigma I've pulled in your changed but I want to do a little more testing today before I merge them.

**Comment by crawford** [comment]
@marineam I've updated the spec to be protocol agnostic. It greatly simplifies the code.

**Comment by sigma** [comment]
@crawford I also have some upcoming changes for reading from OVF environment if present (through the same backdoor, or from the automatically exposed ISO image). I can either open a pull request for your branch, or consider it as a separate change altogether once this one is in place. Any preference?

**Comment by crawford** [comment]
@sigma Let's put that in a separate PR once this is merged.

**Comment by crawford** [comment]
Back peddling a little bit. We need to keep the existing spec, so that's what we'll go with for now. I'm going to follow this up with another PR in a few days to add the new spec side-by-side.

**Comment by marineam** [comment]
[LGTM](http://lgtm.in/i/U88qVjDTp)

**Comment by crawford** [comment]
@sigma I know I said this would be merged, but I just got off the phone with the vendor and they need changes. I hate to do this, but I have to postpone this change a bit longer. The silver lining is that I think we'll be able to switch to the modified scheme (with `coreos.config.url`).

**Comment by crawford** [comment]
I've updated this (once again) with the better version of the spec (we got the OK from the vendor).

File: Documentation/vmware-backdoor.md [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+|:--------------------------------------|:--------------------------------|
+| `hostname`                            | `hostname`                      |
+| `interface.<n>.name`                  | `string`                        |
+| `interface.<n>.mac`                   | `MAC address`                   |
+| `interface.<n>.dhcp`                  | `{"yes", "no"}`                 |
+| `interface.<n>.role`                  | `{"public", "private"}`         |
+| `interface.<n>.ip.<m>.address`        | `CIDR IP address`               |
+| `interface.<n>.route.<l>.gateway`     | `IP address`                    |
+| `interface.<n>.route.<l>.destination` | `CIDR IP address`               |
+| `dns.<x>`                             | `IP address`                    |
```
Comment (marineam) [inline Documentation/vmware-backdoor.md]: lets do `dns.server.<x>` since we may want `dns.domain` or `dns.search.<y>` in the future.

Comment (marineam) [inline Documentation/vmware-backdoor.md]: or not if changes at this point are hard. ;-)

**Comment by marineam** [comment]
LGTM

File: Documentation/vmware-backdoor.md (hunk 2 above)
Comment (crawford) [inline Documentation/vmware-backdoor.md]: Good idea. Done.

**Pull request merged**